samples/gamma-fade/main.c

```diff
@@ -0,0 +1,268 @@
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <math.h>
+#include <hal/video.h>
+#include <hal/debug.h>
+#include <windows.h>
+
+float remap(float in_low,  float in_high,
+            float out_low, float out_high,
+            float value)
+{
+                                 /* [in_low, in_high]       */
+    value -= in_low;             /* [0.0, in_high-in_low]   */
+    value /= in_high - in_low;   /* [0.0, 1.0]              */
+    value *= out_high - out_low; /* [0.0, out_high-out_low] */
+    value += out_low;            /* [out_low, out_high]     */
+    return value;
+}
+
+float linear(float x)
+{
+    return x;
+}
+
+/* This is also known as `smoothstep(0.0, 1.0, x)` */
+float smooth(float x)
+{
+    return x * x * (3.0f - 2.0f * x);
+}
+
+/* This is also known as `lerp(a, b, t)` */
+float mix(float a, float b, float t)
+{
+    return a * (1.0f - t) + b * t;
+}
+
+static void reset_gamma(void)
```

Comment:
This should probably be a standard function

Reply:
I disagree. 

In a real application the user would use their existing gamma control functions.
This could have also been implemented like either of those (to avoid any discontinuity, the real parameters in the caller would be used, with only `t=0.0f`):

```c
// 0% fade to black
mix_fade(0.0f, 0.0f, 0.0f
         0.0f);
```
```c
// 0% fade of linear ramp
cinematic_fade(0.0f, 1.0f, 0.0f, 1.0f,
               0.0f, 1.0f, 0.0f, 1.0f,
               0.0f);
```

I split this because it was more self-documentation, but I can also see how users might not be able to tell what the intention was.


---

I did actually consider this when I wrote the original gamma sample, but here are some arguments:

- If the user doesn't use gamma ramps, then hal/video does this during initialization; so the user doesn't have to worry about their gamma ramps at all. This mimics other features of XVideo like soften or flicker filter.
- If the user does use gamma ramps, then they'll probably have a simple `set_gamma_correction(float gamma)` or they'll abuse it for effects and have a more complicated curve setup to convolve different things `set_gamma_corrected_fade(float gamma, float fade)`.

Providing a ~`XVideoSetGammaCorrection` is pointless and not flexible enough for many users, and we'd have 2 conflicting APIs (low level ramp entry setup, and high level scalar). Auto-convolving those 2 would be confusing and depend on FPU in code that is likely to be ran from DPCs (when handling the gamma ramp update from pbkit in a GPU callback... when we implement those).

Providing ~`XVideoResetGammaRamps` also might lead to accidental calls which break continuity over an animation. This happens if the users gamma function is not close to the linear identity ramp (for example, if their curve generator does primarily do a fade, but also applies gamma correction).

Keeping the XVideo API non-ambigious and lower-level was the best choice (and also mimics how D3D does it).

Reply:
I refactored the code to get rid of this (by ensuring that the function runs for `t=1.0`).

This also avoids an issue where the fade might only have been 99% complete (thereby never reaching full opaqueness). I had previously used `reset_gamma` for the 1% case after the fade, but not for the 99% case between fade-to-color and fade-to-image.